src/include/mpir_datatype.h

```diff
@@ -14,6 +14,64 @@
 #define MPIR_DATATYPE_PAIRTYPE 5
 #define MPIR_DATATYPE_N_PREDEFINED (MPIR_DATATYPE_N_BUILTIN + MPIR_DATATYPE_PAIRTYPE)
 
+/* For internal communication and reduction, many of the builtin-datatypes are
+ * equivallent. Thus, we can make equivallent swap so internally we only need
+ * support a finite essential types */
+
+/* Internal supported types
+ * The handle bit pattern follows the convention of builtin types, except
+ * the last two bytes are type categories rather than indexes.
+ *
+ * Internally we support FIXED, SIGNED, UNSIGNED, FLOAT, and COMPLEX.
+ * Other types can be added when supported, e.g. bfloat16 or logicals that
+ * doesn't follow the C convention of zero/nonzero.
+ */
+#define MPIR_TYPE_FIXED         0x80    /* communication-only, op not supported */
+#define MPIR_TYPE_SIGNED        0x81    /* including C boolean */
+#define MPIR_TYPE_UNSIGNED      0x82
+#define MPIR_TYPE_FLOAT         0x83    /* C native floating point */
+#define MPIR_TYPE_COMPLEX       0x84    /* 2 C native floating point */
+/* Pair types support communication and MPI_MINLOC and MPI_MAXLOC ops. The 1st value type
+ * is SIGNED, UNSIGNED, FLOAT, COMPLEX as defined above. We add a mask for pair types.
+ * E.g. MPI_FLOATINT become MPIR_TYPE_FLOAT | MPIR_TYPE_PAIR_MASK */
+#define MPIR_TYPE_PAIR_MASK      0x10   /* the 2nd type is C int */
+#define MPIR_TYPE_PAIR2_MASK     0x20   /* the 2nd type is the same as 1st type, e.g. MPI_2REAL */
+
+#define MPIR_FIXED0            ((MPI_Datatype)0x4c000080)       /* 0-size, internally equivallent to MPI_DATATYPE_NULL */
```

Comment:
Maybe "BYTE" would be a better name than "FIXED" ?

Reply:
A "BYTE" is always 1 byte :)

Reply:
Sorry, I actually meant `BYTES{0|1|2|4|8|16}`, but OK, I grant it is confusing. 

Just for reference and to provide further alternatives to consideration, NumPy uses "VOID" por opaque sequences of bytes. Such name was most likely inspired from C.

```python
>>> import numpy
>>> numpy.dtype('V4').name
'void32'
>>> numpy.dtype('V4').itemsize
4
>>> numpy.dtype('V4').alignment
1
```